lockc/src/runc.rs

```diff
@@ -0,0 +1,589 @@
+        container_policy_level_POLICY_LEVEL_PRIVILEGED,
+        container_policy_level_POLICY_LEVEL_RESTRICTED,
+    },
+    HashError,
+};
+use lockc_uprobes::{add_container, add_process, delete_container};
+
+// static LABEL_NAMESPACE: &str = "io.kubernetes.pod.namespace";
+static LABEL_POLICY_ENFORCE: &str = "pod-security.kubernetes.io/enforce";
+// static LABEL_POLICY_AUDIT: &str = "pod-security.kubernetes.io/audit";
+// static LABEL_POLICY_WARN: &str = "pod-security.kubernetes.io/warn";
+
+static ANNOTATION_CONTAINERD_LOG_DIRECTORY: &str = "io.kubernetes.cri.sandbox-log-directory";
+static ANNOTATION_CONTAINERD_SANDBOX_ID: &str = "io.kubernetes.cri.sandbox-id";
+
+enum ContainerType {
+    Docker,
+    KubernetesContainerd,
+    Unknown,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Mount {
+    destination: String,
+    r#type: String,
+    source: String,
+    options: Vec<String>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct ContainerConfig {
+    mounts: Vec<Mount>,
+    annotations: Option<HashMap<String, String>>,
+}
+
+#[derive(Error, Debug)]
+pub enum ContainerError {
+    #[error(transparent)]
+    Status(#[from] io::Error),
+
+    #[error(transparent)]
+    Json(#[from] serde_json::Error),
+}
+
+fn container_type_data<P: AsRef<std::path::Path>>(
+    container_bundle: P,
+) -> Result<(ContainerType, std::string::String), ContainerError> {
+    let bundle_path = container_bundle.as_ref();
+    let config_path = bundle_path.join("config.json");
+    let f = fs::File::open(config_path.clone())?;
+    let r = io::BufReader::new(f);
+
+    let config: ContainerConfig = serde_json::from_reader(r)?;
+
+    // Kubernetes
+    if let Some(annotations) = config.annotations {
+        debug!(
+            "detected kubernetes container with bundle {}, config {}",
+            bundle_path.display(),
+            config_path.display(),
+        );
+        // containerd
+        if annotations.contains_key(ANNOTATION_CONTAINERD_LOG_DIRECTORY) {
+            // containerd doesn't expose k8s namespaces directly. They have
+            // to be parsed from the log directory path, where the first
+            // part of the filename is the namespace.
+            let log_directory = &annotations[ANNOTATION_CONTAINERD_LOG_DIRECTORY];
+            debug!(
+                "detected k8s+containerd container with log directory {}",
+                log_directory
+            );
+            let log_path = std::path::PathBuf::from(log_directory);
+            let file_name = log_path.file_name().unwrap().to_str().unwrap();
+            let mut splitter = file_name.split('_');
+            let namespace = splitter.next().unwrap().to_string();
+
+            return Ok((ContainerType::KubernetesContainerd, namespace));
+        // containerd
+        } else if annotations.contains_key(ANNOTATION_CONTAINERD_SANDBOX_ID) {
```

Comment:
For readability, it could be a good idea to split the contains key checks into another function and then return an enum (or other primitive) to `match` off of. This is purely something to think about and I'd be fine with keeping it as is.

Reply:
Good idea, I will try